### PR TITLE
MAINT: Add better error handling in linalg.norm for vectors and clarify it in documentation

### PR DESCRIPTION
**Pull request by Jimver**
From PR #15535 the conclusion was that the documentation for linalg.norm needed to be improved and a better error message needed to be shown when calling `norm` with a vector and `ord='fro'`. This PR enhances the documentation for the linalg.norm() function to explicitly state that the frobenius norm is only defined for two dimensional arrays. See issue #15533 and PR #15535.

### CODE REVIEW

**Comment by Jimver** [comment]
I'd also like to implement a better error message, should I do this in a separate one or this one?

**Comment by rossbar** [comment]
Thanks for your interest @Jimver , this PR should definitely include improved error-handling.

**Comment by Jimver** [comment]
> 
> 
> Thanks for your interest @Jimver , this PR should definitely include improved error-handling.

@rossbar I have added improved error handling to this PR, I also fixed a comment in a test that actually tests for this specific case of calculating the frobenius norm of a vector.

Please let me know if this error handling is good or if I should change the error message.

**Review by rossbar — CHANGES_REQUESTED** [review]
A few initial comments:
 - The explicit case for `ord in ('f', 'fro')` is on the right track, but we should take this opportunity to fixup the rest of the code there. I'm confused about the try-except statement (the `ord + 1` doesn't appear to do anything meaningful except as a *very* roundabout type test?)
 - Note that there is another special kwarg "nuc" that should get the same treatment (currently also seems to be caught by the `try-except`)
 - Make sure any proposed changes obey the 80 char line limit.

I can take a closer look after the next round of changes.

**Comment by Jimver** [comment]
@rossbar I processed the feedback in the latest commit, I use an `isinstance` check if ord is an integer. I also added a separate check for `nuc`.

File: numpy/linalg/linalg.py
Comment (rossbar) [inline numpy/linalg/linalg.py]: Since none of the named kwargs ('fro', 'nuc') are valid for vectors, it seems to me these checks could be consolidated.
```suggestion
        elif ord in ('f', 'fro', 'nuc'):
            raise ValueError(f"Invalid norm order '{ord}' for vectors")
```

File: numpy/linalg/linalg.py
Comment (rossbar) [inline numpy/linalg/linalg.py]: Is this additional condition necessary? Are there cases outside of the 'nuc' and 'fro' for which the `try-catch` you removed are on the path? Other (invalid) str kwargs?

If you can verify that doing so is sufficient for replacing the original `try-except` block that was removed, you could consider further consolidating the above check to:

```python
elif isinstance(ord, str):
    raise ValueError(f"Invalid norm order '{ord}' for vectors")
```

Either way, I think a check based on `numbers.Integral` should be avoided.

File: numpy/linalg/linalg.py
Comment (rossbar) [inline numpy/linalg/linalg.py]: Personally, I don't know whether this extra couple sentences is necessary, but that's an entirely subjective take. If anything, I would try to shorten the additional info while being clear and explicit, maybe something like:

```
Both the Frobenius and nuclear norm orders are only defined for matrices
and raise a ValueError when `x.ndim != 2`.
```

**Review by rossbar — CHANGES_REQUESTED** [review]

File: numpy/linalg/linalg.py
Comment (Jimver) [reply numpy/linalg/linalg.py]: Yeah I agree it is not optimal. I changed it to catch any string as you mentioned in case people have a typo in their `ord` string.

File: numpy/linalg/linalg.py
Comment (Jimver) [reply numpy/linalg/linalg.py]: IMO it is necessary as some people think the Frobenius norm works for vectors according to [Wolfram](https://mathworld.wolfram.com/FrobeniusNorm.html) so this extra clarification makes this more explicitly clear. I changed it to your shorter description to make it more succinct.

File: numpy/linalg/linalg.py
Comment (rossbar) [inline numpy/linalg/linalg.py]:
```suggestion
```
No longer used

File: numpy/linalg/linalg.py
Comment (rossbar) [inline numpy/linalg/linalg.py]:
```suggestion
            return sqrt(add.reduce(s, axis=axis, keepdims=keepdims))
        # None of the str-type keywords for ord ('fro', 'nuc') 
        # are valid for vectors
        elif isinstance(ord, str):
            raise ValueError(f"Invalid norm order '{ord}' for vectors")
```
Since it is not immediately obvious what this branch does, a comment would likely be useful for future readers of the code

File: numpy/linalg/linalg.py
Comment (rossbar) [reply numpy/linalg/linalg.py]: Makes sense to me. One last nitpick - could it be moved so that the line about the nuclear norm is above the new sentence? I think it flows better if it's:
 - explain Frobenius norm
 - explain nuclear norm
 - explain special cases about both fro/nuc for vectors

File: numpy/linalg/linalg.py
Comment (rossbar) [inline numpy/linalg/linalg.py]:
```suggestion
    matrices and raise a ValueError when ``x.ndim != 2``.
```
My fault - I originally suggested markdown syntax, not rST

**Review by rossbar — CHANGES_REQUESTED** [review]
Thanks for the changes - I think this is just about there, one last round of minor tweaks. After that we can ping another reviewer for another set of eyes.

Thanks for adding the additional test as well - that's definitely an improvement from a clarity & coverage standpoint :+1:

File: numpy/linalg/linalg.py
Comment (Jimver) [reply numpy/linalg/linalg.py]: Yes I agree, fixed it just now!

**Comment by Jimver** [comment]
@rossbar I processed your feedback, could you ping other reviewers? I am not so familiar with who I can ping.

**Comment by WarrenWeckesser** [comment]
I'm taking a look now.

**Review by eric-wieser — APPROVED** [review]
Looks good, although PR title should probably be updated to reflect the final patch (which a maintainer can then use in the squashed commit message.

**Review by WarrenWeckesser — APPROVED** [review]
LGTM.

**Comment by WarrenWeckesser** [comment]
The test failures are not related to this PR (something is going wrong with PyPy testing).

**Comment by Jimver** [comment]
> The test failures are not related to this PR (something is going wrong with PyPy testing).

Yes when looking at the logs it is something to do with ubuntu failing to downloading dependencies

**Comment by WarrenWeckesser** [comment]
Do we consider this a fix for gh-15533, so it can be closed when this is merged?

**Comment by WarrenWeckesser** [comment]
Merged.  Thanks @Jimver!  And thanks @rossbar and @eric-wieser for reviews.

**Pull request merged**
